### ISSUE TITLE
bcadmin deferred usage improvements

### ISSUE THREAD

**Issue by jeffallen** [dedis/cothority#1924]
Tasks:

- [x] Update `bcadmin Darc rule` that says -darc has no default
- [x] Discuss upper cased commands like `instID` 
- [x] ~~Create restricted darc with some rules~~
- [x] Hide darc IDs
- [x] Implement a `bcadmin login`
- [x] Deferred spawn: print only once the proposed Tx
- [x] Show args of the instructions in a proposed transaction
- [x] Refactor `--redirect` to `--export` at the contract level
- [x] Make easy to create darc with "N of M" rules
- [x] Make a wrapper to create a deferred Darc

==== Jeff comments ====

1. bcadmin darc rule says -darc has no default, but the default is actually the gnesis darc.
2. --instID is the only flag that's got upper case, can we make flags case insensitive?
3. `bcadmin darc add` cannot set rules at the same time it creates a darc, but this is required when an admin wants to create a darc that will be restricted, since the final users of the darc won't be able to evolve it to add new rules.
4. Never show darc ID, because only the base id is useful in the UI
5. it should be possible to do something like `bcadmin login` that says "from now on, my darc is X and use key Y to sign my requests"

## Deferred usability

1. in spawn: print the instance id of the deferred last, and the proposed txn once.
2. to see the proposed argument, use `bcadmin contract deferred get`, but it needs to show the args, using a heuristic to try to show ASCII if there is enough, see strconv.Quote. Or maybe BasicContract should implement String() and contract writers should be encouraged to override it?

Also: -redirect does not reflect what's happening, and should not be on individual functions. It needs to be on the root of the args tree (`bcadmin -export ...`) so that every call to AddTransactionAndWait in the whole bcadmin can be converted to "deferred capable". In theory, you could make an entire ByzCoin, from the genesis DARC through all the other DARCs and contracts, require a 2-man rule for decentralised authentication. Deferred is not for special contracts, it needs to be possible on all the main contracts, i.e. darc, insecure darc, config, etc.

Consider: easy to use wrappers that help setup a deferred-enabled DARC. Input, list of signers (https://godoc.org/github.com/urfave/cli#StringSliceFlag) and the output is the new DARC id that they will put into the id in the rule of the DARC controlling access to the thing. And then make this thing able to calculate N of M rules (i.e. 2 out of the 4 admins must agree to evolve the roster).

**Comment by ineiti** [comment]
for `5.`, isn't this a bit what `bcadmin link` does by generating a new `bc-xxx.cfg` file?

**Comment by nkcr** [comment]
Tasks:

- [ ] Update `bcadmin Darc rule` that says -darc has no default
- [ ] Discuss upper cased commands like `instID` 
- [ ] Create restricted darc with some rules
- [ ] Hide darc IDs
- [ ] Implement a `bcadmin login`
- [ ] Deferred spawn: print only once the proposed Tx
- [ ] Show args of the instructions in a proposed transaction
- [ ] Refactor `--redirect` to `--export` at the contract level
- [ ] Make easy to create darc with "N of M" rules

**Comment by nkcr** [comment]
@jeffallen For point 4. we directly use the `String()` method of Darc in `cothority/darc/darc.go` to print it. This method is potentially used elsewhere, should we update it?
